contracts/peg/NonCustodialPSM.sol

```diff
@@ -0,0 +1,463 @@
+
+/// @notice Peg Stability Module that holds no funds.
+/// On a mint, it transfers all proceeds to a PCV Deposit
+/// When funds are needed for a redemption, they are simply pulled from the PCV Deposit
+contract NonCustodialPSM is
+    OracleRef,
+    RateLimited,
+    ReentrancyGuard,
+    INonCustodialPSM
+{
+    using Decimal for Decimal.D256;
+    using SafeCast for *;
+    using SafeERC20 for IERC20;
+
+    /// @notice the fee in basis points for selling an asset into Fei
+    uint256 public override mintFeeBasisPoints;
+
+    /// @notice the fee in basis points for buying the asset for Fei
+    uint256 public override redeemFeeBasisPoints;
+
+    /// @notice the PCV deposit target to deposit and withdraw from
+    IPCVDeposit public override pcvDeposit;
+
+    /// @notice the token this PSM will exchange for Fei
+    /// Must be a stable token pegged to $1
+    IERC20 public immutable override underlyingToken;
+
+    /// @notice Rate Limited Minter contract that will be called when Fei needs to be minted
+    GlobalRateLimitedMinter public override rateLimitedMinter;
+
+    /// @notice the max mint and redeem fee in basis points
+    /// Governance cannot change the maximum fee
+    uint256 public immutable override MAX_FEE = 300;
+
+    /// @notice boolean switch that indicates whether redeeming is paused
+    bool public redeemPaused;
+
+    /// @notice boolean switch that indicates whether minting is paused
+    bool public mintPaused;
+
+    /// @notice struct for passing constructor parameters related to OracleRef
+    struct OracleParams {
+        address coreAddress;
+        address oracleAddress;
+        address backupOracle;
+        int256 decimalsNormalizer;
+    }
+
+    /// @notice struct for passing constructor parameters related to MultiRateLimited
+    struct RateLimitedParams {
+        uint256 maxRateLimitPerSecond;
+        uint256 rateLimitPerSecond;
+        uint256 bufferCap;
+    }
+
+    /// @notice struct for passing constructor parameters related to the non custodial PSM
+    struct PSMParams {
+        uint256 mintFeeBasisPoints;
+        uint256 redeemFeeBasisPoints;
+        IERC20 underlyingToken;
+        IPCVDeposit pcvDeposit;
+        GlobalRateLimitedMinter rateLimitedMinter;
+    }
+
+    /// @notice construct the non custodial PSM. Structs are used to prevent stack too deep errors
+    /// @param params oracle ref constructor data
+    /// @param rateLimitedParams rate limited constructor data
+    /// @param psmParams non custodial PSM constructor data
+    constructor(
+        OracleParams memory params,
+        RateLimitedParams memory rateLimitedParams,
+        PSMParams memory psmParams
+    )
+        OracleRef(
+            params.coreAddress,
+            params.oracleAddress,
+            params.backupOracle,
+            params.decimalsNormalizer,
+            true /// hardcode doInvert to true to allow swaps to work correctly
```

Comment:
You may not need doInvert set to true for this PR. Consider changing this to a param in the OracleParams struct